lib/query_tracker.rb

```diff
@@ -0,0 +1,30 @@
+require 'pg_query'
+
+# Tracks queries by table
+# Intended for tests, not for production
+class QueryTracker
+  # @yield block to track queries during
+  # @return [Hash<Symbol, Array<Array(:Symbol, String)>>] a hash of
+  #   {table_name: [[:select, "query"]]}
+  def self.track
+    queries = Hash.new { |h, k| h[k] = [] }
+
+    subscriber = ActiveSupport::Notifications.
+      subscribe('sql.active_record') do |_name, _start, _finish, _id, payload|
+        sql = payload[:sql]
+
+        action = sql.split(' ').first.downcase.to_sym
+        tables = PgQuery.parse(sql).tables.map(&:to_sym)
+
+        tables.each do |table|
+          queries[table] << [action, sql]
+        end
```

Comment:
Maybe a YAGNI, or also general unease of seeing raw SQL tracked (even for tests), and since we're not testing the behavior, should we just concern ourselves with count for now?

Roughly:

```suggestion
        tables.each do |table|
          queries[table]++
        end
```

Reply:
I can see about trying to log the file + line to also help track things down

Reply:
For that benefit, is there a way we can have it so the [assertion below](https://github.com/18F/identity-idp/blob/c8c2a51e995ff9abf1cd2ab0b988dd88216d5cfc/spec/features/openid_connect/openid_connect_spec.rb#L625) would log the inspected result of the array, rather than the count?

Reply:
The queries we have are very long (because we removed `SELECT *` by default in #5342, SELECT * can be considered an antipattern)... so I don't think that would be a very useful assertion to have the full query bodies there?

For a simple `ServiceProvider.find_by(issuer: 'aaa')` we get:

```
  ServiceProvider Load (1.8ms)  SELECT "service_providers"."id", "service_providers"."issuer", "service_providers"."friendly_name", "service_providers"."description", "service_providers"."metadata_url", "service_providers"."acs_url", "service_providers"."assertion_consumer_logout_service_url", "service_providers"."logo", "service_providers"."signature", "service_providers"."block_encryption", "service_providers"."sp_initiated_login_url", "service_providers"."return_to_sp_url", "service_providers"."attribute_bundle", "service_providers"."created_at", "service_providers"."updated_at", "service_providers"."active", "service_providers"."approved", "service_providers"."native", "service_providers"."redirect_uris", "service_providers"."agency_id", "service_providers"."failure_to_proof_url", "service_providers"."ial", "service_providers"."piv_cac", "service_providers"."piv_cac_scoped_by_email", "service_providers"."pkce", "service_providers"."push_notification_url", "service_providers"."help_text", "service_providers"."allow_prompt_login", "service_providers"."ial2_quota", "service_providers"."signed_response_message_requested", "service_providers"."liveness_checking_required", "service_providers"."remote_logo_key", "service_providers"."launch_date", "service_providers"."iaa", "service_providers"."iaa_start_date", "service_providers"."iaa_end_date", "service_providers"."app_id", "service_providers"."default_aal", "service_providers"."certs", "service_providers"."email_nameid_format_allowed", "service_providers"."use_legacy_name_id_behavior" FROM "service_providers" WHERE "service_providers"."issuer" = $1 LIMIT $2  [["issuer", "aaa"], ["LIMIT", 1]]
```

Reply:
So, when you say having the full query will make it easier to figure out where it came from, from what perspective would you be debugging those? Are you thinking a `pry` breakpoint or something?

Reply:
My thinking is, I'd run a test with the query tracker, and expect less than X queries

If the test failed and there were more, I'd drop in a `pp` to just dump the hash or a `pry` if the print was too much

I think I'd like knowing the file + line to see where it came from is what I'd use to track down usage, so I just updated in 4f5175c85d342bd432a7e58077d4fa9c3b2988eb to use name + line

Reply:
wfm 👍

Reply:
Another alternative interface would be to return a class that has more useful methods on it, or even just define a custom matcher that could have a useful error message

```ruby
queries = QueryTracker.track { ... }

# using a hash
expect(queries).to have_queried(:users).no_more_than(3).times
# expected no more than 3 queries, got 4
# - foobar_controller.rb:6 (:create)
# - foobar_controller.rb:7 (:insert)
# - bar_baz_controller.rb:100 (:insert)
# - bar_baz_controller.rb:101 (:insert)
```